src/genai_detection/pipelines/data_processing/nodes.py

```diff
@@ -17,52 +17,113 @@ def transform_raw_image_data(raw_image_dataset):
     num_features = model.fc.in_features
     model.fc = torch.nn.Linear(num_features, 1)
     return model
 
+
 model_initialization_node = node(
                 func=get_model,
                 inputs=None,
                 outputs="model",
                 name="model_initialization",
 )
 
-def train_model(model, dataset):
-    batch_size = TRAIN_CONFIG["batch_size"]
+
+def initialize_wandb_run(config):
+    """Initialize a W&B run."""
+    return create_run(config=config)
+
+
+initialize_wandb_run_node = node(
+                func=initialize_wandb_run,
+                inputs="config",
+                outputs="wandb_run",
+                name="wandb_run_initialization",
+)
+
+
+def train_model(model, dataset, config, wandb_run):
+    """Train the model on the given dataset based on config yaml."""
+    batch_size = config["batch_size"]
     dataloader = torch.utils.data.DataLoader(dataset, batch_size=batch_size, shuffle=True)
-    for epoch in range(TRAIN_CONFIG["epochs"]):
+    for epoch in range(config["epochs"]):
         for i, data in enumerate(dataloader, 0):
             inputs, labels = data
             optimizer = torch.optim.SGD(
                 model.parameters(),
-                lr=TRAIN_CONFIG["learning_rate"],
-                momentum=TRAIN_CONFIG["momentum"],
+                lr=config["learning_rate"],
+                momentum=config["momentum"],
                 )
             optimizer.zero_grad()
             outputs = torch.sigmoid(model(inputs))
             loss_fn = torch.nn.BCELoss()
             loss = loss_fn(outputs, labels.reshape(-1, 1).to(torch.float))
             loss.backward()
             optimizer.step()
-            LOGGER.info(f"Epoch: {epoch}, Loss: {loss.item()}")
+            log_metrics(wandb_run,
+                        {
+                            "loss": loss.item(),
+                            "epoch": epoch,
+                        },
+                        step=(epoch+1)*(i+1))
 
     return model
 
+
 train_model_node = node(
                 func=train_model,
-                inputs=["model", "preprocessed_data"],
+                inputs=["model", "preprocessed_data", "config", "wandb_run"],
                 outputs="trained_model",
                 name="model_training",
 )
+
+
+def save_and_upload_model(model, wandb_run, config):
+    """Save the trained model and upload it to W&B."""
+    save_checkpoint(wandb_run, model, config['model_name'])
+    return wandb_run
+
+save_and_upload_model_node = node(
+                func=save_and_upload_model,
+                inputs=["trained_model", "wandb_run", "config"],
+                outputs="finished_wandb_run",
+                name="model_saving_and_uploading",
+)
+
+def finish_wandb_run(wandb_run) -> None:
+    wandb_run.finish()
+
+finish_wandb_run_node = node(
+                func=finish_wandb_run,
+                inputs="finished_wandb_run",
+                outputs=None,
+                name="wandb_run_finishing",
+)
```

Comment:
Tu wystarczy jako `func` podać `wandb_run.finish`:
`func=wanb_run.finish` (bez nawiasów) i powinno działać tak samo

Reply:
tl;dr: budujemy tutaj graf obliczeniowy kedro i możemy posługiwać się tylko funkcjami bezstanowymi oraz nazwami własnymi przekazywanych obiektów. Nie mozemy podać tego tak imperatywnie.

`wandb_run` nie będzie istniał w tej przestrzeni zmiennych, bo istnieje tylko jako parametr wywołania tej funkcji. Musimy podać to poprzez `inputs`, gdzie też nie mamy bezpośrednio dostępu do tego obiektu, a jedynie podajemy jego nazwę `finished_wandb_run`, która też jest celowo inna od poprzednich nazw wandb_runu, żeby ten node nie zakończył runu nim skończy się node treningowy.